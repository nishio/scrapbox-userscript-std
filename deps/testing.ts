export * from "https://deno.land/std@0.217.0/testing/asserts.ts";
export * from "https://deno.land/std@0.217.0/testing/snapshot.ts";
