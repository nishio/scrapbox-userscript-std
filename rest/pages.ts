import type {
  ErrorLike,
  NotFoundError,
  NotLoggedInError,
  NotMemberError,
  Page,
  PageList,
} from "@cosense/types/rest";
import { cookie } from "./auth.ts";
import { parseHTTPError } from "./parseHTTPError.ts";
import { encodeTitleURI } from "../title.ts";
import { type BaseOptions, setDefaults } from "./util.ts";
import {
  andThenAsyncForResult,
  mapAsyncForResult,
  mapErrAsyncForResult,
  type Result,
} from "option-t/plain_result";
import { unwrapOrForMaybe } from "option-t/maybe";
import { type HTTPError, responseIntoResult } from "./responseIntoResult.ts";
import type { AbortError, NetworkError } from "./robustFetch.ts";

/** Options for `getPage()` */
export interface GetPageOption extends BaseOptions {
  /** use `followRename` */
  followRename?: boolean;

  /** project ids to get External links */
  projects?: string[];
}

export interface TooLongURIError extends ErrorLike {
  name: "TooLongURIError";
}

const getPage_toRequest: GetPage["toRequest"] = (
  project,
  title,
  options,
) => {
  const { sid, hostName, followRename, projects } = setDefaults(options ?? {});
  const params = new URLSearchParams();
  params.append("followRename", `${followRename ?? true}`);
  for (const id of projects ?? []) {
    params.append("projects", id);
  }

  return new Request(
    `https://${hostName}/api/pages/${project}/${
      encodeTitleURI(title)
    }?${params}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );
};

const getPage_fromResponse: GetPage["fromResponse"] = async (res) =>
  mapErrAsyncForResult(
    await mapAsyncForResult(
      responseIntoResult(res),
      (res) => res.json() as Promise<Page>,
    ),
    async (
      error,
    ) => {
      if (error.response.status === 414) {
        return {
          name: "TooLongURIError",
          message: "project ids may be too much.",
        };
      }

      return unwrapOrForMaybe<
        NotFoundError | NotLoggedInError | NotMemberError | HTTPError
      >(
        await parseHTTPError(error, [
          "NotFoundError",
          "NotLoggedInError",
          "NotMemberError",
        ]),
        error,
      );
    },
  );

export interface GetPage {
  /** /api/pages/:project/:title の要求を組み立てる
   *
   * @param project 取得したいページのproject名
   * @param title 取得したいページのtitle 大文字小文字は問わない
   * @param options オプション
   * @return request
   */
  toRequest: (
    project: string,
    title: string,
    options?: GetPageOption,
  ) => Request;

  /** 帰ってきた応答からページのJSONデータを取得する
   *
   * @param res 応答
   * @return ページのJSONデータ
   */
  fromResponse: (res: Response) => Promise<
    Result<
      Page,
      | NotFoundError
      | NotLoggedInError
      | NotMemberError
      | TooLongURIError
      | HTTPError
    >
  >;

  (project: string, title: string, options?: GetPageOption): Promise<
    Result<
      Page,
      | NotFoundError
      | NotLoggedInError
      | NotMemberError
      | TooLongURIError
      | HTTPError
      | NetworkError
      | AbortError
    >
  >;
}

/** 指定したページのJSONデータを取得する
 *
 * @param project 取得したいページのproject名
 * @param title 取得したいページのtitle 大文字小文字は問わない
 * @param options オプション
 */
export const getPage: GetPage = async (
  project,
  title,
  options,
) =>
  andThenAsyncForResult<
    Response,
    Page,
    | NotFoundError
    | NotLoggedInError
    | NotMemberError
    | TooLongURIError
    | HTTPError
    | NetworkError
    | AbortError
  >(
    await setDefaults(options ?? {}).fetch(
      getPage_toRequest(project, title, options),
    ),
    (input) => getPage_fromResponse(input),
  );

getPage.toRequest = getPage_toRequest;
getPage.fromResponse = getPage_fromResponse;

/** Options for `listPages()` */
export interface ListPagesOption extends BaseOptions {
  /** the sort of page list to return
   *
   * @default "updated"
   */
  sort?:
    | "updatedWithMe"
    | "updated"
    | "created"
    | "accessed"
    | "pageRank"
    | "linked"
    | "views"
    | "title";
  /** the index getting page list from
   *
   * @default 0
   */
  skip?: number;
  /** threshold of the length of page list
   *
   * @default 100
   */
  limit?: number;
}

export interface ListPages {
  /** /api/pages/:project の要求を組み立てる
   *
   * @param project 取得したいページのproject名
   * @param options オプション
   * @return request
   */
  toRequest: (
    project: string,
    options?: ListPagesOption,
  ) => Request;

  /** 帰ってきた応答からページのJSONデータを取得する
   *
   * @param res 応答
   * @return ページのJSONデータ
   */
  fromResponse: (res: Response) => Promise<
    Result<
      PageList,
      NotFoundError | NotLoggedInError | NotMemberError | HTTPError
    >
  >;

  (project: string, options?: ListPagesOption): Promise<
    Result<
      PageList,
      | NotFoundError
      | NotLoggedInError
      | NotMemberError
      | NetworkError
      | AbortError
      | HTTPError
    >
  >;
}

const listPages_toRequest: ListPages["toRequest"] = (project, options) => {
  const { sid, hostName, sort, limit, skip } = setDefaults(
    options ?? {},
  );
  const params = new URLSearchParams();
  if (sort !== undefined) params.append("sort", sort);
  if (limit !== undefined) params.append("limit", `${limit}`);
  if (skip !== undefined) params.append("skip", `${skip}`);

  return new Request(
    `https://${hostName}/api/pages/${project}?${params}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );
};

const listPages_fromResponse: ListPages["fromResponse"] = async (res) =>
  mapErrAsyncForResult(
    await mapAsyncForResult(
      responseIntoResult(res),
      (res) => res.json() as Promise<PageList>,
    ),
    async (error) =>
      unwrapOrForMaybe<
        NotFoundError | NotLoggedInError | NotMemberError | HTTPError
      >(
        await parseHTTPError(error, [
          "NotFoundError",
          "NotLoggedInError",
          "NotMemberError",
        ]),
        error,
      ),
  );

/** 指定したprojectのページを一覧する
 *
 * @param project 一覧したいproject
 * @param options オプション 取得範囲や並び順を決める
 */
export const listPages: ListPages = async (
  project,
  options?,
) =>
  andThenAsyncForResult<
    Response,
    PageList,
    | NotFoundError
    | NotLoggedInError
    | NotMemberError
    | NetworkError
    | AbortError
    | HTTPError
  >(
    await setDefaults(options ?? {})?.fetch(
      listPages_toRequest(project, options),
    ),
    listPages_fromResponse,
  );

listPages.toRequest = listPages_toRequest;
listPages.fromResponse = listPages_fromResponse;
