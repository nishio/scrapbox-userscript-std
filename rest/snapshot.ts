import type {
  ErrorLike,
  NotFoundError,
  NotLoggedInError,
  NotMemberError,
  PageSnapshotList,
  PageSnapshotResult,
} from "@cosense/types/rest";
import { cookie } from "./auth.ts";
import { type BaseOptions, setDefaults } from "./options.ts";
import { parseHTTPError } from "./parseHTTPError.ts";
import {
  isErr,
  mapAsyncForResult,
  mapErrAsyncForResult,
  type Result,
  unwrapOk,
} from "option-t/plain_result";
import { type HTTPError, responseIntoResult } from "./responseIntoResult.ts";
import type { FetchError } from "./mod.ts";

/** Error that occurs when an invalid `timestampId` is provided to {@linkcode getSnapshot} */
export interface InvalidPageSnapshotIdError extends ErrorLike {
  name: "InvalidPageSnapshotIdError";
}

export type SnapshotError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | InvalidPageSnapshotIdError
  | HTTPError;

/** get a page snapshot
 *
 * @param options `connect.sid` etc.
 */
export const getSnapshot = async (
  project: string,
  pageId: string,
  timestampId: string,
  options?: BaseOptions,
): Promise<Result<PageSnapshotResult, SnapshotError | FetchError>> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}/${timestampId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  if (isErr(res)) return res;

  return mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(unwrapOk(res)),
      async (error) =>
        error.response.status === 422
          ? {
            name: "InvalidPageSnapshotIdError",
            message: await error.response.text(),
          }
          : (await parseHTTPError(error, [
            "NotFoundError",
            "NotLoggedInError",
            "NotMemberError",
          ])) ?? error,
    ),
    (res) => res.json() as Promise<PageSnapshotResult>,
  );
};

export type SnapshotTimestampIdsError =
  | NotFoundError
  | NotLoggedInError
  | NotMemberError
  | HTTPError;

/**
 * Retrieves the timestamp IDs for a specific page in a project.
 *
 * @param project - The name of the project.
 * @param pageId - The ID of the page.
 * @param options - Optional configuration options.
 * @returns A promise that resolves to a {@linkcode Result} object containing the page snapshot list if successful,
 * or an error if the request fails.
 */
export const getTimestampIds = async (
  project: string,
  pageId: string,
  options?: BaseOptions,
): Promise<
  Result<PageSnapshotList, SnapshotTimestampIdsError | FetchError>
> => {
  const { sid, hostName, fetch } = setDefaults(options ?? {});

  const req = new Request(
    `https://${hostName}/api/page-snapshots/${project}/${pageId}`,
    sid ? { headers: { Cookie: cookie(sid) } } : undefined,
  );

  const res = await fetch(req);
  if (isErr(res)) return res;

  return mapAsyncForResult(
    await mapErrAsyncForResult(
      responseIntoResult(unwrapOk(res)),
      async (error) =>
        (await parseHTTPError(error, [
          "NotFoundError",
          "NotLoggedInError",
          "NotMemberError",
        ])) ?? error,
    ),
    (res) => res.json() as Promise<PageSnapshotList>,
  );
};
