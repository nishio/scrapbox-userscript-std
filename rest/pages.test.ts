import { getPage, listPages } from "./pages.ts";
import { assertSnapshot } from "@std/testing/snapshot";

/** Test suite for page retrieval functionality */
Deno.test("getPage", async (t) => { // Tests page fetching with various options
  // Test fetching a page with rename following enabled
  await assertSnapshot(
    t,
    getPage.toRequest("takker", "テストページ", { followRename: true }),
  );
});
/** Test suite for page listing functionality */
Deno.test("listPages", async (t) => { // Tests page listing with sorting options
  await assertSnapshot(
    t,
    listPages.toRequest("takker", { sort: "updated" }),
  );
});
