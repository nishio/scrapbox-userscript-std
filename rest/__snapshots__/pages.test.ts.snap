export const snapshot = {};

snapshot[`getPage 1`] = `
Request {
  bodyUsed: false,
  headers: Headers {},
  method: "GET",
  redirect: "follow",
  url: "https://scrapbox.io/api/pages/takker/%E3%83%86%E3%82%B9%E3%83%88%E3%83%9A%E3%83%BC%E3%82%B8?followRe"... 9 more characters
}
`;

snapshot[`listPages 1`] = `
Request {
  bodyUsed: false,
  headers: Headers {},
  method: "GET",
  redirect: "follow",
  url: "https://scrapbox.io/api/pages/takker?sort=updated"
}
`;
